proposals/yield-in-try-catch.md

```diff
@@ -334,52 +362,135 @@ bool MoveNext()
 
 ## Considerations
 
-### yield inside catch with nested finally
-
-[catch-finally]: #yield-inside-catch-with-nested-finally
-
-The `yield` statement cannot be reasonbly supported inside a `catch` blocks with a nested `finally` due to the behavior of the `Dispose` method. It is possible that a `catch` block will run as part of executing a `finally` in the `Dispose` method.
+### yield inside finally
 
-For example consider the following:
+Consideration was given to supporting `yield` inside a `finally` block but this creates significant semantic challenges. Consider the following code as an example:
 
 ```csharp
-var e = InCatchFinally();
+var e = M().GetEnumerator();
 e.MoveNext();
 e.Dispose();
 
-IEnumerable<int> InCatchFinally()
+void M()
+{
+    try
+    {
+        yield return 42;
+    }
+    finally
+    {
+        yield return 13;
+    }
+}
+```
+
+The `finally` block would be executed in `Dispose` which means the language has to decide on the behavior of `yield` during `Dispose`. That could be modeled as:
+
+1. Throw an exception when `yield` is encountered in `Dispose`.
+2. Silently ignore the value and continue executing the `finally` block without suspend.
+
+Neither of these seem like desirable outcomes and as such `yield` will not be allowed in `finally` blocks.
+
+## Open Issues
+
+### Preserve catch in Dispose paths
+
+The `Dispose` method could include mirroring both `catch` and `finally` blocks. This would allow the `catch` block to be executed in `Dispose` when `finally` blocks threw an exceptoin. The approach for this would be to do the followng.
+
+For every `catch` block where the `try` has a nested `try / finally` with `yield`:
+
+1. The contents of the `catch` block will be generated into a parameterless method on the iterator with a `void` return.
+2. The contents of the `when` clause will be generated into a parameterless method on the iterator with a `bool` return.
+3. The `catch` block will be replaced with a call to the generated method in `MoveNext`.
+4. The `when` clause will be replaced with a call to the generated method in `MoveNext`.
+5. The `Dispose` method will mirror `catch` and `finally` blocks in the same way it mirrors `finally` blocks today and dispatch to the appropriate method
```

Comment:
Could we also do codegen similar to async iterators, i.e., share MoveNext and Dispose implementations?

Reply:
That is definitely an option, essentially changing iterators to favor the async iterators code gen approach. I _think_ that would be fine as I can't see any technical reason why iterators have to be done that way but we'd want to vet it out a bit